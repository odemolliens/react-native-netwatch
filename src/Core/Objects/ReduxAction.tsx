import { ILog, LogType } from '../../types';

interface IAction {
  type: string;
  payload: any;
}

export class ReduxAction implements ILog {
  _id: number = -1;
  type: LogType = 'REDUX';
  startTime: number = Date.now();
  stringifiedAction: string = '';
  extra: string = '';
  action: IAction = { type: '__ERROR:UNDEFINED__', payload: '' };

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ReduxAction;
