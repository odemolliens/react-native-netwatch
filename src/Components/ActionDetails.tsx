import * as React from 'react';
import { useContext } from 'react';
import { View, StyleSheet } from 'react-native';
import { Subheading } from 'react-native-paper';
import { reduxTag } from './Status';
import { getTime, getShortDate, isLongText, addEllipsis } from '../Utils/helpers';
import { Text, Title } from './Text';
import { ThemeContext } from '../Theme';
import ReduxAction from '../Core/Objects/ReduxAction';
import { ViewMoreButton } from './ViewMoreButton';

export interface IProps {
  testId?: string;
  item: ReduxAction;
  reduxAction: any;
  onPressViewMore: (value: boolean) => void;
}

export const ActionDetails: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <View style={{ flex: 1, width: '100%' }}>
      {/* REDUX + REDUX TAG */}
      <View style={[styles.line]}>
        <Title style={[{ marginRight: 6 }, { color: theme.textColorOne }]}>REDUX</Title>
        {reduxTag()}
      </View>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Started at : </Text>
        <Text>{`${getShortDate(props.item.startTime)} - ${getTime(props.item.startTime)}`}</Text>
      </View>

      <Subheading
        style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
      >
        Content
      </Subheading>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Type</Text>
      </View>
      <View style={[styles.line]}>
        <Text>{props.item.action.type}</Text>
      </View>

      <View style={[styles.line]}>
        {props.reduxAction.label && (
          <Text style={{ color: theme.textColorFour }}>{`${props.reduxAction.label
            .charAt(0)
            .toUpperCase()}${props.reduxAction.label.slice(1)} :`}</Text>
        )}
      </View>
      <View style={[styles.line]}>
        <View style={[styles.attribtuesContainer, { flex: 1, paddingHorizontal: 0 }]}>
          {props.reduxAction.payload && <Text>{addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))}</Text>}
          {isLongText(addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))) && (
            <View style={{ alignItems: 'flex-end' }}>
              <ViewMoreButton onPress={() => props.onPressViewMore(true)} />
            </View>
          )}
        </View>
      </View>
    </View>
  );
};

export default ActionDetails;

const styles = StyleSheet.create({
  line: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    marginBottom: 10,
    alignItems: 'center',
  },

  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },

  attribtuesContainer: {
    justifyContent: 'flex-start',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
});
