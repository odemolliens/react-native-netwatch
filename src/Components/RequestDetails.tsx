import * as React from 'react';
import { useContext, useState } from 'react';
import { View, StyleSheet, Image } from 'react-native';
import { Subheading } from 'react-native-paper';
import {
  getTime,
  getShortDate,
  isLongText,
  addEllipsis,
  getStatus,
  duration,
  getGeneralElementsAsArray,
  getRequestHeadersElementsAsArray,
  getResponseHeadersElementsAsArray,
  excludedAttributesForExport,
} from '../Utils/helpers';
import { Text, Title } from './Text';
import { tag } from './Status';
import { ThemeContext } from '../Theme';
import { NRequest } from '../Core/Objects/NRequest';
import { RNRequest } from '../Core/Objects/RNRequest';
import { EnumStatus } from '../types';
import url from 'url';
// @ts-ignore
import JSONDetails from './JSONDetails';
import { ViewMoreButton } from './ViewMoreButton';
import Clipboard from '@react-native-clipboard/clipboard';
import ClipboardButton from './ClipboardButton';

export interface IProps {
  testId?: string;
  item: NRequest | RNRequest;
  onPressViewMoreRequest: (value: boolean) => void;
  onPressViewMoreResponse: (value: boolean) => void;
  setSnackBarMessage: (value: string) => void;
  setSnackBarVisibility: (value: boolean) => void;
}

// This component is specif to request (React-Native or Native).
// If you need the componant which handle the Redux action, see
// ActionDetails.tsx
export const RequestDetails: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);
  const [onErrorImage, setOnErrorImage] = useState<boolean>(false);

  let _color: string = theme.reduxColor;
  const _temp = getStatus(props.item.status);
  if (_temp === EnumStatus.Success) _color = theme.successColor;
  if (_temp === EnumStatus.Warning) _color = theme.warningColor;
  if (_temp === EnumStatus.Failed) _color = theme.failedColor;
  const urlObject = url.parse(props.item.url, true);
  const hostname = urlObject.host || '';
  const _params = Object.entries(urlObject.query);
  const _generalElements = getGeneralElementsAsArray(props.item);
  const _requestHeadersElements = getRequestHeadersElementsAsArray(props.item);
  const _responseHeadersElements = getResponseHeadersElementsAsArray(props.item);

  // The elements are generated here for _generalElements, _requestHeadersElements and _responseHeadersElements
  // There is a filter with excludedAttributesForExport to avoid duplicate elements
  // For example, status is present in the root of props.item but also in _requestHeadersElements
  // that's why status is listed in excludedAttributesForExport
  const _renderItems = (listOfItems: Array<[string, any]>) => {
    return listOfItems
      .filter((item: Array<string>) => item.length > 1) // To be sure that item has at least two element
      .filter((item: Array<string>) => !excludedAttributesForExport.includes(item[0]))
      .filter((item: Array<string>) => item[1] && item[1].length > 0)
      .map((item: Array<string>, index: number) => {
        return (
          <View style={styles.itemContainer} key={index}>
            {typeof item[1] === 'string' && item[1].startsWith('data:image/') ? (
              _renderImage(item[1])
            ) : (
              <>
                <View style={styles.attribute}>
                  <Text style={[{ color: theme.textColorFour }]}>{item[0]}</Text>
                  {item[0] === 'url' && _copyClipbutton(item[1], 'URL has been copied to clipboard')}
                </View>
                <Text style={[{ width: '100%' }, { color: theme.textColorOne }]}>{item[1]}</Text>
              </>
            )}
          </View>
        );
      });
  };

  // If an item contains a base64 image. This one will be displayed directly instead of plain text
  const _renderImage = (source: string) => {
    return (
      <>
        {onErrorImage ? (
          _renderErrorMessage()
        ) : (
          <View style={{ minHeight: 100, justifyContent: 'center', alignItems: 'center' }}>
            <Image
              source={{ uri: source }}
              style={{ width: '80%', height: '100%' }}
              resizeMode="contain"
              onError={() => setOnErrorImage(true)}
            />
          </View>
        )}
      </>
    );
  };

  const _copyToClipboard = (text: string): void => {
    if (typeof text === 'string') Clipboard.setString(text);
  };

  const _copyClipbutton = (text: string = '', toastMessage: string = '') => {
    return (
      <ClipboardButton
        onPress={() => {
          _copyToClipboard(text);
          props.setSnackBarMessage(toastMessage);
          props.setSnackBarVisibility(true);
        }}
      />
    );
  };

  const _renderErrorMessage = () => {
    return <Text>An error occur, cannot load the image</Text>;
  };

  return (
    <View style={{ flex: 1, width: '100%' }}>
      {/* METHOD + STATUS CODE */}
      <View style={[styles.line]}>
        <Title style={[{ marginRight: 6 }, { color: theme.textColorOne }]}>{props.item.method}</Title>
        {tag(_color, props.item.status.toString())}
      </View>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Hostname : </Text>
        <Text>{hostname}</Text>
      </View>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Started at : </Text>
        <Text>{`${getShortDate(props.item.startTime)} - ${getTime(props.item.startTime)}   ( ${duration(
          props.item.startTime,
          props.item.endTime,
        )}ms )`}</Text>
      </View>

      {typeof props.item.timeout === 'number' && (
        <View style={[styles.line]}>
          <Text style={{ color: theme.textColorFour }}>Timeout : </Text>
          <Text>{props.item?.timeout.toString()}</Text>
        </View>
      )}

      {_generalElements.length > 0 && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Request Info
          </Subheading>
          {_renderItems(_generalElements)}
        </>
      )}

      {_params.length > 0 && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Request Params
          </Subheading>
          {_renderItems(_params)}
        </>
      )}

      {_requestHeadersElements.length > 0 && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Request Headers
          </Subheading>
          {_renderItems(_requestHeadersElements)}
        </>
      )}

      {props.item.dataSent?.length > 0 && props.item.dataSent !== 'undefined' && props.item.dataSent !== 'null' && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Request Data
          </Subheading>
          <View style={styles.attribtuesContainer}>
            <Text>{addEllipsis(props.item?.dataSent)}</Text>
            {isLongText(props.item?.dataSent) && (
              <View style={{ alignItems: 'flex-end' }}>
                <ViewMoreButton onPress={() => props.onPressViewMoreRequest(true)} />
              </View>
            )}
          </View>
        </>
      )}

      {_responseHeadersElements.length > 0 && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Response Headers
          </Subheading>
          {_renderItems(_responseHeadersElements)}
        </>
      )}

      {props.item?.response?.length > 0 && props.item.response !== 'undefined' && props.item.response !== 'null' && (
        <>
          <Subheading
            style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
          >
            Response Body
          </Subheading>
          <View style={{ flexDirection: 'row-reverse', paddingHorizontal: 16 }}>
            {_copyClipbutton(props.item?.response, 'Response has been copied to clipboard')}
          </View>
          <View style={styles.attribtuesContainer}>
            <Text>{addEllipsis(props.item?.response)}</Text>
            {isLongText(props.item?.response) && (
              <View style={{ alignItems: 'flex-end' }}>
                <ViewMoreButton onPress={() => props.onPressViewMoreResponse(true)} />
              </View>
            )}
          </View>
        </>
      )}
    </View>
  );
};

export default RequestDetails;

const styles = StyleSheet.create({
  attribute: {
    marginBottom: 10,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },

  itemContainer: {
    paddingHorizontal: 16,
    marginBottom: 10,
  },

  line: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    marginBottom: 10,
    alignItems: 'center',
  },

  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },

  attribtuesContainer: {
    justifyContent: 'flex-start',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },

  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
