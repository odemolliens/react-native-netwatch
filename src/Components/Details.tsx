import * as React from 'react';
import { useContext, useState } from 'react';
import { StyleSheet, View, ScrollView, Alert, TouchableOpacity } from 'react-native';
import Share from 'react-native-share';
import { Appbar, Snackbar } from 'react-native-paper';
import RNRequest from '../Core/Objects/RNRequest';
import NRequest from '../Core/Objects/NRequest';
import ReduxAction from '../Core/Objects/ReduxAction';
import FeatherIcon from 'react-native-vector-icons/Feather';
import { ThemeContext } from '../Theme';
import { ActionDetails } from './ActionDetails';
import JSONDetails from './JSONDetails';
import {
  getGeneralElementsAsArray,
  getRequestHeadersElementsAsArray,
  getResponseHeadersElementsAsArray,
  formatSharedMessage,
} from '../Utils/helpers';
import { RequestDetails } from './RequestDetails';
import { ILog } from '../types';

export interface IProps {
  testId?: string;
  onPressBack: (showDetails: boolean) => void;
  item: ILog;
}

export const Details: React.FC<IProps> = props => {
  const theme = useContext(ThemeContext);
  const [snackBarVisibility, setSnackBarVisibility] = useState<boolean>(false);
  const [snackBarMessage, setSnackBarMessage] = useState<string>('');
  const [showJSONResponseDetails, setShowJSONResponseDetails] = useState<boolean>(false);
  const [showJSONRequestDetails, setShowJSONRequestDetails] = useState<boolean>(false);
  const [showJSONActionDetails, setShowJSONActionDetails] = useState<boolean>(false);
  let _content = null;
  let _action: any = () => {};

  // _onShareRequest and _onShareReduxAction open the share panel. The text format is
  // plain text
  const _onShareRequest = async (item: RNRequest | NRequest): Promise<void> => {
    try {
      await Share.open({
        message: formatSharedMessage(
          getGeneralElementsAsArray(item),
          getRequestHeadersElementsAsArray(item),
          item.dataSent,
          getResponseHeadersElementsAsArray(item),
          item.response,
        ),
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  const _onShareReduxAction = async (item: ReduxAction): Promise<void> => {
    const _type = item.action.type.toUpperCase();
    const _payload = JSON.stringify(item.action.payload, null, 2);
    try {
      await Share.open({
        message: `${_type}\n${_payload}`,
      });
    } catch (error) {
      Alert.alert('Error', error.message);
    }
  };

  // To avoid heavy library such as react-navigation, the screen with JSON tree
  // is displayed by using the showJSON***Details state for the
  // corresponding instance (Action or request). This screen can have five (5) differents views

  // 1. We want to display the action Redux in a JSON Tree
  if (showJSONActionDetails && props.item instanceof ReduxAction) {
    return (
      <JSONDetails
        title="Action details"
        onPressBack={() => setShowJSONActionDetails(false)}
        data={props.item?.action}
      />
    );
  }

  // 2. We want to display the a request RESPONSE BODY only in a JSON Tree
  if (showJSONResponseDetails && (props.item instanceof RNRequest || props.item instanceof NRequest)) {
    return (
      <JSONDetails
        title="Response details"
        onPressBack={() => setShowJSONResponseDetails(false)}
        data={props.item?.response}
      />
    );
  }

  // 3. We want to display the a request REQUEST BODY only in a JSON Tree
  if (showJSONRequestDetails && (props.item instanceof RNRequest || props.item instanceof NRequest)) {
    return (
      <JSONDetails
        title="Request details"
        onPressBack={() => setShowJSONRequestDetails(false)}
        data={props.item?.dataSent}
      />
    );
  }

  // 4. We want to display ALL INFORMATIONS in a redux action
  // At this time (for point 4 and 5), we store the result in _content variable
  // to be showed later.
  if (props.item instanceof ReduxAction) {
    // props.item.action should only contains 2 elements, a type and a payload (not necessary called payload).
    // In consequence, if it's not the type, that could be the payload
    const _infos = Object.entries(props.item.action).filter(value => value[0] !== 'type');
    const _reduxAction = {
      label: (_infos.length > 0 && _infos[0][0]) || undefined,
      payload: (_infos.length > 0 && _infos[0][1]) || undefined,
    };

    _action = () => _onShareReduxAction(props.item as ReduxAction);
    _content = (
      <ActionDetails
        item={props.item}
        reduxAction={_reduxAction}
        onPressViewMore={() => setShowJSONActionDetails(true)}
      />
    );
  }

  // 5. We want to display ALL INFORMATIONS in a request (React-Native or Native)
  if (props.item instanceof RNRequest || props.item instanceof NRequest) {
    _action = () => _onShareRequest(props.item as RNRequest | NRequest);
    _content = (
      <RequestDetails
        item={props.item}
        onPressViewMoreRequest={() => setShowJSONRequestDetails(true)}
        onPressViewMoreResponse={() => setShowJSONResponseDetails(true)}
        setSnackBarMessage={setSnackBarMessage}
        setSnackBarVisibility={setSnackBarVisibility}
      />
    );
  }

  // Appbar header is repeated here cause we use the absolute position in the style
  // Put this directly in the index.tsx cause that the Appbar will be added two times.
  return (
    <View style={styles.container}>
      <Appbar.Header style={[styles.header, { backgroundColor: theme.secondaryDarkColor }]}>
        <TouchableOpacity
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => props.onPressBack(false)}
          testID="buttonBackToMainScreen"
        >
          <FeatherIcon name="arrow-left" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
        <Appbar.Content color={theme.primaryColor} title="Netwatch" titleStyle={{ fontSize: 18 }} />
        <TouchableOpacity
          testID="buttonShare"
          style={[styles.button, { borderLeftWidth: 0 }]}
          onPress={() => _action()}
        >
          <FeatherIcon name="download" color={theme.textColorOne} size={24} />
        </TouchableOpacity>
      </Appbar.Header>

      <ScrollView style={{ backgroundColor: theme.secondaryColor }} contentContainerStyle={styles.scrollview}>
        {_content}
      </ScrollView>
      <Snackbar
        visible={snackBarVisibility}
        onDismiss={() => setSnackBarVisibility(false)}
        duration={3000}
        style={{ backgroundColor: theme.textColorOne }}
        theme={{
          colors: {
            surface: theme.textColorFour,
          },
        }}
      >
        {snackBarMessage}
      </Snackbar>
    </View>
  );
};

const styles = StyleSheet.create({
  header: {
    elevation: 0,
    shadowOffset: {
      width: 0,
      height: 0,
    },
    shadowOpacity: 0,
  },

  container: {
    position: 'absolute',
    height: '100%',
    width: '100%',
    top: 0,
    left: 0,
  },

  scrollview: {
    paddingTop: 26,
    paddingBottom: 20,
  },

  button: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 56,
    width: 56,
    borderLeftWidth: 1,
  },
});
