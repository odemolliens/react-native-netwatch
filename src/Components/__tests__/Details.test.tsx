import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { Details, IProps } from '../Details';
import { RNRequest } from '../../Core/Objects/RNRequest';
import { NRequest } from '../../Core/Objects/NRequest';
import { ReduxAction } from '../../Core/Objects/ReduxAction';
import Share from 'react-native-share';
import ActionDetails from '../ActionDetails';
import RequestDetails from '../RequestDetails';

describe('Details test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  const back = jest.fn();
  const setShowJSONResponseDetails = jest.fn();
  const setShowJSONRequestDetails = jest.fn();
  const setShowJSONActionDetails = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  beforeEach(() => {
    jest.mock('react', () => ({
      ...jest.requireActual('react'),
      useState: jest.fn(),
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Test Redux item', () => {
    it('should render Redux action properly', () => {
      givenProps(mockActionWithLongPayload);
      givenComponent();
      expect(component).toMatchSnapshot();
    });

    it('should share actions', () => {
      givenProps(mockAction);
      givenComponent();
      component.find(`[testID="buttonShare"]`).simulate('press');
      expect(Share.open).toHaveBeenCalledTimes(1);
    });

    it('should called setShowJSONActionDetails', () => {
      const useStateMock: any = (showJSONActionDetails: any) => [showJSONActionDetails, setShowJSONActionDetails];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(mockActionWithLongPayload);
      givenComponent();
      component.find(ActionDetails).invoke('onPressViewMore')(true);
      component.update();
      expect(setShowJSONActionDetails).toHaveBeenCalledTimes(1);
    });
  });

  describe('Test Request item', () => {
    it('should share requests', () => {
      givenProps(mockNRequestWithLongBody);
      givenComponent();
      component.find(`[testID="buttonShare"]`).simulate('press');
      expect(Share.open).toHaveBeenCalledTimes(1);
    });

    it('should render properly with request & press backbutton', () => {
      givenProps(mockNRequestWithLongBody);
      givenComponent();
      expect(component).toMatchSnapshot();
      whenPressingButton('buttonBackToMainScreen');
      expect(back).toHaveBeenCalledTimes(1);
    });
  });

  describe('Tests useState Action', () => {
    it('should show JSON Details screen with actions', () => {
      const useStateMock: any = (showJSONActionDetails: any) => [true, setShowJSONActionDetails];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(mockActionWithLongPayload);
      givenComponent();
      expect(component).toMatchInlineSnapshot(`
        <JSONDetails
          data={
            Object {
              "payload": "{\\"data\\":\\"This is a test with a long payload, more than 100\\",\\"next\\":\\"characters to check if the viewmore button appears correctly.\\",\\"explain\\":\\"In fact, we need to have at least 100 characters to show this button\\",\\"and\\":\\" test if we can press the button\\"}",
              "type": "__ERROR:UNDEFINED__",
            }
          }
          onPressBack={[Function]}
          title="Action details"
        />
      `);
    });
  });

  describe('Tests useState Request', () => {
    it('should called setShowJSONResponseDetails', () => {
      const useStateMock: any = (showJSONResponseDetails: any) => [showJSONResponseDetails, setShowJSONResponseDetails];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(mockRequestWithLongResponse);
      givenComponent();
      component.find(RequestDetails).invoke('onPressViewMoreResponse')(true);
      expect(setShowJSONResponseDetails).toHaveBeenCalledTimes(1);
    });

    it('should show JSON Details screen with response', () => {
      const useStateMock: any = (showJSONResponseDetails: any) => [true, setShowJSONResponseDetails];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(mockRequestWithLongResponse);
      givenComponent();
      expect(component).toMatchInlineSnapshot(`
        <JSONDetails
          data="{\\"data\\":\\"This is a test with a long response, more than 100\\",\\"next\\":\\"characters to check if the viewmore button appears correctly.\\",\\"explain\\":\\"In fact, we need to have at least 100 characters to show this button\\",\\"and\\":\\" test if we can press the button\\"}"
          onPressBack={[Function]}
          title="Response details"
        />
      `);
    });
    it('should called setShowJSONRequestDetails', () => {
      const useStateMock: any = (showJSONRequestDetails: any) => [showJSONRequestDetails, setShowJSONRequestDetails];
      jest.spyOn(React, 'useState').mockImplementation(useStateMock);
      givenProps(mockNRequestWithLongBody);
      givenComponent();
      component.find(RequestDetails).invoke('onPressViewMoreRequest')(true);
      expect(setShowJSONRequestDetails).toHaveBeenCalledTimes(1);
    });
  });

  // GIVEN
  function givenComponent() {
    component = shallow(<Details {...props} />);
  }

  function givenProps(item?: RNRequest | ReduxAction) {
    props = {
      testId: 'testId',
      onPressBack: back,
      item,
    };
  }

  // UTILITIES

  function whenPressingButton(testId: string) {
    component.find(`[testID="${testId}"]`).simulate('press');
  }

  const mockAction: ReduxAction = new ReduxAction({
    _id: 73,
    startTime: 100,
    stringifiedAction: "{ type: '__ERROR:UNDEFINED__', payload: '' }",
    type: 'REDUX',
    action: { type: '__ERROR:UNDEFINED__', payload: '' },
  });

  const mockActionWithLongPayload: ReduxAction = new ReduxAction({
    _id: 73,
    startTime: 100,
    stringifiedAction: "{ type: '__ERROR:UNDEFINED__', payload: '' }",
    type: 'REDUX',
    action: {
      type: '__ERROR:UNDEFINED__',
      payload: JSON.stringify({
        data: 'This is a test with a long payload, more than 100',
        next: 'characters to check if the viewmore button appears correctly.',
        explain: 'In fact, we need to have at least 100 characters to show this button',
        and: ' test if we can press the button',
      }),
    },
  });

  const mockNRequestWithLongBody: NRequest = new NRequest({
    _id: 75,
    dataSent: JSON.stringify({
      data: 'This is a test with a long body, more than 100',
      next: 'characters to check if the viewmore button appears correctly.',
      explain: 'In fact, we need to have at least 100 characters to show this button',
      and: ' test if we can press the button',
    }),
    endTime: 1613477575757,
    method: 'GET',
    readyState: 4,
    requestHeaders: {
      'Content-Type': 'application/json',
    },
    response: 'response',
    responseContentType: 'application/json',
    responseHeaders: {
      'Content-Length': '0',
      'Content-Type': 'application/json; charset=UTF-8',
      Date: 'Tue, 16 Feb 2021 12:12:55 GMT',
      'Sozu-Id': '51989c0c-ebe7-4574-913d-443477875da7',
    },
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 200,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  });

  const mockRequestWithLongResponse: RNRequest = new RNRequest({
    _id: 75,
    dataSent: 'body',
    endTime: 1613477575757,
    method: 'GET',
    readyState: 4,
    requestHeaders: {
      'Content-Type': 'application/json',
    },
    response: JSON.stringify({
      data: 'This is a test with a long response, more than 100',
      next: 'characters to check if the viewmore button appears correctly.',
      explain: 'In fact, we need to have at least 100 characters to show this button',
      and: ' test if we can press the button',
    }),
    responseContentType: 'application/json',
    responseHeaders: {
      'Content-Length': '0',
      'Content-Type': 'application/json; charset=UTF-8',
      Date: 'Tue, 16 Feb 2021 12:12:55 GMT',
      'Sozu-Id': '51989c0c-ebe7-4574-913d-443477875da7',
    },
    responseSize: 0,
    responseType: 'blob',
    responseURL: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
    startTime: 1613477574742,
    status: 200,
    timeout: 0,
    type: 'RNR',
    url: 'https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9',
  });
});
