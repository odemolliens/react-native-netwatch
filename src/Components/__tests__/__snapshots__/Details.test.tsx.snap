// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details test suite Test Redux item should render Redux action properly 1`] = `
<Component
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
    }
  }
>
  <Header
    style={
      Array [
        Object {
          "elevation": 0,
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
        },
        Object {
          "backgroundColor": "#111827",
        },
      ]
    }
  >
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
      testID="buttonBackToMainScreen"
    >
      <Icon
        color="#F9FAFB"
        name="arrow-left"
        size={24}
      />
    </ForwardRef>
    <Content
      color="#06B6D4"
      title="Netwatch"
      titleStyle={
        Object {
          "fontSize": 18,
        }
      }
    />
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
      testID="buttonShare"
    >
      <Icon
        color="#F9FAFB"
        name="download"
        size={24}
      />
    </ForwardRef>
  </Header>
  <Component
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingTop": 26,
      }
    }
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
  >
    <ActionDetails
      item={
        ReduxAction {
          "_id": 73,
          "action": Object {
            "payload": "{\\"data\\":\\"This is a test with a long payload, more than 100\\",\\"next\\":\\"characters to check if the viewmore button appears correctly.\\",\\"explain\\":\\"In fact, we need to have at least 100 characters to show this button\\",\\"and\\":\\" test if we can press the button\\"}",
            "type": "__ERROR:UNDEFINED__",
          },
          "extra": "",
          "startTime": 100,
          "stringifiedAction": "{ type: '__ERROR:UNDEFINED__', payload: '' }",
          "type": "REDUX",
        }
      }
      onPressViewMore={[Function]}
      reduxAction={
        Object {
          "label": "payload",
          "payload": "{\\"data\\":\\"This is a test with a long payload, more than 100\\",\\"next\\":\\"characters to check if the viewmore button appears correctly.\\",\\"explain\\":\\"In fact, we need to have at least 100 characters to show this button\\",\\"and\\":\\" test if we can press the button\\"}",
        }
      }
    />
  </Component>
  <Snackbar
    duration={3000}
    onDismiss={[Function]}
    style={
      Object {
        "backgroundColor": "#F9FAFB",
      }
    }
    theme={
      Object {
        "colors": Object {
          "surface": "#6B7280",
        },
      }
    }
    visible={false}
  />
</Component>
`;

exports[`Details test suite Test Request item should render properly with request & press backbutton 1`] = `
<Component
  style={
    Object {
      "height": "100%",
      "left": 0,
      "position": "absolute",
      "top": 0,
      "width": "100%",
    }
  }
>
  <Header
    style={
      Array [
        Object {
          "elevation": 0,
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0,
        },
        Object {
          "backgroundColor": "#111827",
        },
      ]
    }
  >
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
      testID="buttonBackToMainScreen"
    >
      <Icon
        color="#F9FAFB"
        name="arrow-left"
        size={24}
      />
    </ForwardRef>
    <Content
      color="#06B6D4"
      title="Netwatch"
      titleStyle={
        Object {
          "fontSize": 18,
        }
      }
    />
    <ForwardRef
      onPress={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderLeftWidth": 1,
            "height": 56,
            "justifyContent": "center",
            "width": 56,
          },
          Object {
            "borderLeftWidth": 0,
          },
        ]
      }
      testID="buttonShare"
    >
      <Icon
        color="#F9FAFB"
        name="download"
        size={24}
      />
    </ForwardRef>
  </Header>
  <Component
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingTop": 26,
      }
    }
    style={
      Object {
        "backgroundColor": "#1F2937",
      }
    }
  >
    <RequestDetails
      item={
        NRequest {
          "_id": 75,
          "dataSent": "{\\"data\\":\\"This is a test with a long body, more than 100\\",\\"next\\":\\"characters to check if the viewmore button appears correctly.\\",\\"explain\\":\\"In fact, we need to have at least 100 characters to show this button\\",\\"and\\":\\" test if we can press the button\\"}",
          "endTime": 1613477575757,
          "method": "GET",
          "readyState": 4,
          "requestHeaders": Object {
            "Content-Type": "application/json",
          },
          "response": "response",
          "responseContentType": "application/json",
          "responseHeaders": Object {
            "Content-Length": "0",
            "Content-Type": "application/json; charset=UTF-8",
            "Date": "Tue, 16 Feb 2021 12:12:55 GMT",
            "Sozu-Id": "51989c0c-ebe7-4574-913d-443477875da7",
          },
          "responseSize": 0,
          "responseType": "blob",
          "responseURL": "https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9",
          "shortUrl": "",
          "startTime": 1613477574742,
          "status": 200,
          "timeout": 0,
          "type": "RNR",
          "url": "https://run.mocky.io/v3/1a2d092a-42b2-4a89-a44f-267935dc13e9",
        }
      }
      onPressViewMoreRequest={[Function]}
      onPressViewMoreResponse={[Function]}
      setSnackBarMessage={[Function]}
      setSnackBarVisibility={[Function]}
    />
  </Component>
  <Snackbar
    duration={3000}
    onDismiss={[Function]}
    style={
      Object {
        "backgroundColor": "#F9FAFB",
      }
    }
    theme={
      Object {
        "colors": Object {
          "surface": "#6B7280",
        },
      }
    }
    visible={false}
  />
</Component>
`;
